### DISCUSSION TITLE
Large Number of Deletions

### DISCUSSION THREAD

**Discussion by picheny-nyu** [NVIDIA-NeMo/NeMo#7083] [Q&A]
I have some noisy low quality data; I put the data through a few Nemo speech recognition models and see a lot of deletions (relative to, say, wav2vec and whisper). Is there some setting I can adjust to reduce the number of deletions? I am using the models out of the box without any fine tuning.

Thanks
Michael Picheny

**Comment by titu1994** [comment]
Which model are you using from NeMo? You can try to use Conformer or Fast Conformer transducer or hybrid model to get a baseline

**Reply by picheny-nyu** [reply]
I tried both the conformer transducer large and extra large. The extra large had many more deletions than the large.

**Reply by titu1994** [reply]
Also to make sure, youre converting the audio to monochannel 16khz data right ?

Hmm you can try CTC models or fast conformer hybrid xlarge, but depending on audio quality it might not beat whisper. I'm surprised Wav2Vec can do well against Nemo models though.

**Reply by titu1994** [reply]
CTC - https://huggingface.co/nvidia/stt_en_conformer_ctc_large 

Fast Conformer Hybrid - https://huggingface.co/nvidia/stt_en_fastconformer_transducer_xlarge

**Reply by picheny-nyu** [reply]
Yes it is mono channel 16 kHz. It is low snr and not much info above 4khz, also many short utterances it is challenging data. 

The wav2vec2 results are after fine tuning. Fine tuning on whisper not doing much but out of box it is very good. 

If you could point me to an easy to use fine tuning example I would be glad to try it. I have a small amount of labelled data and also data to train an external lm...

Thanks for your quick reponses also.

**Reply by picheny-nyu** [reply]
Fastconformer had fewer deletions than conformer but still a lot. CTC had significantly fewer deletions than Fastconformer (but did have a few). Again, pointers to fine tuning scripts. would be really helpful!!

**Reply by picheny-nyu** [reply]
Sent.     wds. Acc.   Sub.   Del.    Ins.   WER. SER
ctc-l:           | Sum/Avg|  406    2602 | 63.5   22.9   13.6    2.1   38.5   77.8 |
rnnt-fast-xl | Sum/Avg|  406    2602 | 57.6   14.4   28.0    1.5   43.9   77.6 |
rnnt-xl         | Sum/Avg|  406    2602 | 58.7   13.0   28.3    1.1   42.4   77.6 |
rnnt-l           | Sum/Avg|  406    2602 | 61.5   15.9   22.6    1.2   39.7   80.8 |

For comparison Whisper-medium yields:
medium      | Sum/Avg|  406     2602 | 74.1   17.9     7.9    3.7   29.6    68.0 |

**Reply by titu1994** [reply]
Whisper has been trained on a lot of private data with music and noise, so it is more robust to severely degraded audio quality. It seems CTC models are close to the whisper model, and to improve their wer you can try using beam search with them. 

Finetuning tutorial is available here - https://colab.research.google.com/github/NVIDIA/NeMo/blob/stable/tutorials/asr/ASR_CTC_Language_Finetuning.ipynb

You can write a short script based on that to do the finetuning with your requirements.

**Reply by titu1994** [reply]
Also, fast conformer hybrid can be run in CTC mode - you can use the transcribe speech.py in examples with decoder_type='ctc' here - https://github.com/NVIDIA/NeMo/blob/main/examples/asr/transcribe_speech.py#L157

Or manually load the model and call `model.change_decoding_strategy(None, decoder_type="ctc")`

**Reply by picheny-nyu** [reply]
Thanks so much. I suspect that rnnt fine tuning would also help a lot and I would be happy to try and report back, would you have a similar pointer? Thanks again so much for your help!

**Reply by titu1994** [reply]
Right RNNT might do fine with some finetuning. 

RNNT finetuning, and hybrid model finetuning is almost exactly following the same recipe as the CTC finetuning. You should be able to follow those steps for all 3. If you need any help, you can ask here or in a issue and we will help. 

You'll have to call model.change_vocabulary() with the new tokenizer (if you use a new Tokenizer, otherwise it's better to continue to use the inbuilt one).

**Reply by picheny-nyu** [reply]
The fine-tuning tutorial you point me to is for Japanese. Since my data is English I don't need most of the steps. Would you happened to have a pointer to script that just fine tunes an English model from English data? It would save me some time filtering out the unnecessary steps. If not, I can struggle through it. Thanks again!
